Add backward support for rudimentary NestedTensor.sum(dim)

mikaylagawarecki opened the pull request:
Per offline discussion, this will be updated to use expand once expand semantics for nested tensor have been fleshed out.

Next steps will be to add support for other features for forward sum mentioned on #82387 and likewise update the backward


Stack from [ghstack](https://github.com/ezyang/ghstack) (oldest at bottom):
* __->__ #82625
* #83356

mikaylagawarecki committed:
Add backward support for rudimentary NestedTensor.sum(dim)

[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"




[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"




[ghstack-poisoned]

mikaylagawarecki commented on tools/autograd/derivatives.yaml:
```diff
@@ -1532,11 +1532,11 @@
+  self: sum_backward_symint(grad, self.sym_sizes(), dim, keepdim)
   result: auto_linear
 
 - name: sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
```
@albanD  do you have any thoughts on what is the best way to register a NestedTensor kernel for a backward formula that accepts `self.sizes()`? It seems that this is the case for most of the reductions :( 

Right now I'm doing something super hacky where I add a kwarg self to sum_backward and pass a dummy in place of sizes which is probably really bad. I recall Driss mentioned something about editing VariableTypeManual.cpp but you mentioned yesterday that we probably shouldn't do this, is there any other way around this?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1532,11 +1532,11 @@
 
 - name: sum.dim_IntList(Tensor self, int[1]? dim, bool keepdim=False, *, ScalarType? dtype=None) -> Tensor
-  self: sum_backward(grad, self.sizes(), dim, keepdim)
+  self: sum_backward(grad, IntArrayRef({-1}), dim, keepdim, self)
```
The problem is that this is not OK to do. The reason is that if you use `self` here, then the autograd has to save the full Tensor of size `self` and thus this will increase memory usage for all users.

mikaylagawarecki replied:
@albanD completely agree that what I'm doing here is horrible, was just hacking together something to try to get it to compile! If I understand correctly from talking to Driss before, (please correct me if I'm wrong) the scheme in the linked google doc was not implemented for now. I think I heard during the meeting yesterday that it was possible to workaround the `self.sizes()` issue for NestedTensor in VariableTypeManual.cpp though, would you be willing to point me to an example of how this could be done?

albanD commented:
This was already discussed in https://docs.google.com/document/d/1dVMH33IK2C7n3Ee8LBtfWw2i8MzxkAIkvuUo_hZY63k/edit#
Any reason for not going with that is proposed there?

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"




[ghstack-poisoned]

bdhirsh commented on aten/src/ATen/native/native_functions.yaml:
```diff
@@ -4771,6 +4771,11 @@
   device_check: NoCheck   # TensorIterator
 
+# TODO: this function will be replaced once nested expand semantics have been settled on
+- func: _nested_sum_backward(Tensor grad, Tensor self, int[1]? dim, bool keepdim=False) -> Tensor
+  dispatch:
+    NestedTensorCPU: NestedTensor_sum_dim_backward_CPU
```
tiny nit: there isn't really any enforcement about naming conventions in the dispatch table, but maybe name the kernel `_nested_sum_backward_cpu`? (i.e. `{foo}_cpu` for the cpu kernel for an op `{foo}`.

albanD replied:
nit: the usual (from the rest of this file) naming convention for this would be `_nested_sum_backward_cpu`.
I know this is not the current convention for a few nested function but it would be good to align the two no?

albanD commented on tools/autograd/derivatives.yaml:
```diff
@@ -1544,8 +1544,13 @@
-  result: auto_linear
+  dispatch:
+    Default:
+      self: sum_backward(grad, self.sizes(), dim, keepdim)
```
nit: Could you add the following test to TestAutogradMultipleDispatch please?
```python
    def test_sum_not_saving(self, device):
        def foo(x):
            # Don't modify the input inplace
            x = x.clone()
            res = x.sum()
            x.add_(1)
            return res

        inp = torch.rand(2, device=device, requires_grad=True)
        # sum's input is not saved for regular Tensors
        foo(inp).backward()

        # sum's input is saved for nested Tensors
        nt = torch.nested_tensor([torch.rand(2), torch.rand(2)], device=device).requires_grad_()
        with self.assertRaisesRegex(RuntimeError, "has been modified inplace"):
            foo(nt).backward()
```

(I can't run it locally, so there might be a bug there. But something in the same spirit.

albanD commented on test/test_nestedtensor.py:
```diff
@@ -616,17 +616,32 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
-            nt2 = torch.nested_tensor(ub2)
-            self.assertEqual(nt, nt2)
+            if requires_grad:
+                nt.requires_grad_(requires_grad)
```
Why isn't `requires_grad` a kwarg for the `torch.nested_tensor()` constructor? Is that constructor supposed to be differentiable?

mikaylagawarecki replied:
I believe it is supposed to be differentiable, will look into adding this!

albanD replied:
If it is differentiable (which looks like it is), you shouldn't add that.

albanD commented on test/test_nestedtensor.py:
```diff
@@ -616,17 +616,32 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
+            self.assertEqual(nt_sum, torch.nested_tensor(ub2_sum))
+            # test backward
+            if requires_grad:
+                # generate gradient tensor that has the same size as the output
```
It would be nice to have `torch.randn_like(nt_sum)` :D

mikaylagawarecki replied:
Could I hold off on this until Driss implements empty_like in #83140, I think that the implementation for randn_like might be similar

albanD replied:
Of course!

btw driss is using `nt = self.random_nt(device, dtype, 4, (4, 4))` in a few of his tests. Maybe that is a good replacement?

In any case, this is a small detail that can be done later.

albanD replied:
Ho you're already using that now :D

albanD commented on test/test_nestedtensor.py:
```diff
@@ -616,17 +616,32 @@ def test_nested_tensor_mul_in_place(self, device, dtype):
 
         for ntensors, max_sizes in params:
-            test_sum(self.random_nt(device, dtype, ntensors, max_sizes), len(max_sizes))
+            test_sum(device, dtype, ntensors, max_sizes, len(max_sizes), requires_grad=True)
```
If requires_grad is always True, no need for it to be a kwarg?

mikaylagawarecki commented on tools/autograd/derivatives.yaml:
```diff
@@ -1544,8 +1544,13 @@
+    AutogradNestedTensor:
+      # TODO: replace this function once semantics for nested tensor expand have been settled on
+      self: _nested_sum_backward(grad, self, dim, keepdim)
 
```
Hm I was under the impression that it was not supported from[ this TORCH_CHECK ](https://github.com/pytorch/pytorch/blob/master/torch/csrc/autograd/forward_grad.cpp#L20) :b

albanD replied:
Ho wow, well I guess it is not :)

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"

Per offline discussion, this will be updated to use expand once expand semantics for nested tensor have been fleshed out.





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"

Per offline discussion, this will be updated to use expand once expand semantics for nested tensor have been fleshed out.





[ghstack-poisoned]

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"

Per offline discussion, this will be updated to use expand once expand semantics for nested tensor have been fleshed out.





[ghstack-poisoned]

albanD reviewed (approved):
SGTM
I didn't look into details into the kernel though (but I guess it's fine as it's going to be removed soon).

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot rebase -s

mikaylagawarecki committed:
Update on "Add backward support for rudimentary NestedTensor.sum(dim)"

Per offline discussion, this will be updated to use expand once expand semantics for nested tensor have been fleshed out.

Next steps will be to add support for other features for forward sum mentioned on #82387 and likewise update the backward





[ghstack-poisoned]

mikaylagawarecki commented:
@pytorchbot merge

mikaylagawarecki commented:
@pytorchbot merge
